core/src/main/scala/kafka/server/KafkaRaftServer.scala

```diff
@@ -102,6 +106,8 @@ class KafkaRaftServer(
       threadNamePrefix,
       controllerQuorumVotersFuture,
       KafkaRaftServer.configSchema,
+      raftApiVersions,
+      Some(bootstrapMetadata)
```

Comment:
That's an interesting point. For the broker, on first start, it will start up without a `metadata.version`. It won't be until the bootstrapped version is written by the controller and replicated to the broker that BrokerMetadataPublisher will update the FinalizedFeatureCache. 

This is probably okay though, since brokers start up in a fenced state in KRaft anyways. Since the `metadata.version` bootstrapping is done in the leader election callback on the controller, we should have the FeatureLevelRecord written out before any broker registrations occur. I think this should avoid any issues with a broker becoming unfenced before knowing what the `metadata.version` is.

Does that make sense? Or did I misunderstand your question :)

Reply:
Yeah, that makes sense. But I want to clarify one thing here:
> we should have the FeatureLevelRecord written out before any broker registrations occur

The broker will register itself regardless of whether FeatureLevelRecord is written, but only set wantfence=false in the HeartbeatReq after FeatureLevelRecord has been published.